actionpack/lib/action_dispatch/middleware/cookies.rb

```diff
@@ -599,9 +599,16 @@ class EncryptedKeyRotatingCookieJar < AbstractCookieJar # :nodoc:
       def initialize(parent_jar)
         super
 
-        key_len = ActiveSupport::MessageEncryptor.key_len(encrypted_cookie_cipher)
-        secret = request.key_generator.generate_key(request.authenticated_encrypted_cookie_salt, key_len)
-        @encryptor = ActiveSupport::MessageEncryptor.new(secret, cipher: encrypted_cookie_cipher, serializer: SERIALIZER)
+        if request.use_authenticated_cookie_encryption
```

Comment:
Thought `upgrade_legacy_hmac_aes_cbc_cookies?` was would handle this? I don't understand the need for this `if` at all, and I certainly wouldn't just reading the code 😊

Reply:
I feel like the expectation of setting `action_dispatch.use_authenticated_cookie_encryption` to `false` would be encrypted cookies use CBC encryption.

Reply:
I really don't see who would assign this to `false`. It's an upgrade option.

Apps generated on 5.2 default to true and we hide the config in https://github.com/rails/rails/blob/6bc2ac64cea6e148c84356b998314b61a9405754/railties/lib/rails/application/configuration.rb#L98 — they're not meant to switch it off.

Apps upgrading from 5.1 get a `new_framework_defaults_5_2.rb` containing a commented out `# config.use_authenticated_cookie_encryption = true`. In which case we fallback to adding a legacy encryptor.

Are you saying that we default  — no matter the value of that config — generate cookies that can't be read by Rails 5.1 apps? E.g. making it impossible for apps to rollback in case of a haywire 5.2 upgrade?

Reply:
Would users upgrading from an older version to 5.2 ever set this to `false`?

Rolling back is a different question, I think. If a cookie has already been rewritten using GCM and the version is then rolled back an earlier version, there's nothing in place in 5.1 (or the earlier version) to rotate "downward" from GCM to CBC. This could also be problematic if people were to use different versions in production...

Reply:
From what I gather here, my concern is that for 5.1 upgrading users even if the config is commented out, `# config.use_authenticated_cookie_encryption = true`, that we're still going to be upgrading their cookies behind their backs.

Once users then comment the config in there's no going back to a previous Rails version. I'm okay with that as long as we're clear about that in the `new_framework_defaults_5_2.rb` file.

Reply:
> From what I gather here, my concern is that for 5.1 upgrading users even if the config is commented out, # config.use_authenticated_cookie_encryption = true, that we're still going to be upgrading their cookies behind their backs.

I believe this is correct. For this case, we'd be writing cookies with GCM and not even trying to to do upgrade rotation from CBC to GCM. Thus, I think if `action_dispatch.use_authenticated_cookie_encryption` is `false`, we should just use CBC for encrypted cookies.